### PR TITLE
feat: pass known file sizes to filesystem in Python

### PR DESCRIPTION
**Pull request by eeroel**
# Description
Currently the Filesystem implementation always makes a HEAD request when opening a file, to determine the file size. The proposed change is to read the file sizes from the delta log instead, and to pass them down to the `open_input_file` call, eliminating the HEAD request.

# Related Issue(s)

# Documentation

### CODE REVIEW

File: python/deltalake/fs.py [hunk 1]
```diff
@@ -12,23 +12,47 @@ class DeltaStorageHandler(DeltaFileSystemHandler, FileSystemHandler):
...
+
+    def __init__(
+        self,
+        table_uri: str,
+        storage_options: Optional[Dict[str, str]] = None,
+        known_sizes: Optional[Dict[str, int]] = None,
+    ):
+        if known_sizes:
+            self.known_sizes = known_sizes
+        return
```
Comment (wjones127) [inline python/deltalake/fs.py:35]: Could we push the `known_sizes` mapping down into the Rust `DeltaFileSystemHandler`? Then we won't need to change the signatures of `open_input_file` or `open_input_stream`.

Comment (eeroel) [reply python/deltalake/fs.py]: Good point, fixed this 👍

**Comment by eeroel** [comment]
Fixed formatting issue and rebased on main

**Comment by eeroel** [comment]
Fixed remaining issue with type annotation

**Review by wjones127 — APPROVED** [review]
This looks good. I appreciate the tests :)

**Pull request merged**